File: problems/LeetCode/backtrack/permute.py
def permute(self, nums: List[int]) -> List[List[int]]:

		n = len(nums)
		path = [0] * n
		on_path = [False] * n
		ans = []

		def dfs(i:int) -> None:
				if i == n:
						ans.append(path.copy())
						return
				
				for j in range(n):
						if not on_path[j]:
								path[i] = nums[j]
								on_path[j] = True
								dfs(i+1)
								on_path[j] = False
		dfs(0)
		return ans

File: problems/LeetCode/TopInterview150/permute.ts
function permute(nums: number[]): number[][] {
  const ans: number[][] = [];

  const backtrack = (prem: Set<number>) => {
    if (prem.size === nums.length) {
      ans.push([...prem]);
      return;
    }

    for (const num of nums) {
      if (prem.has(num)) continue;
      prem.add(num);
      backtrack(prem);
      prem.delete(num);
    }
  };

  backtrack(new Set());
  return ans;
}

function permute(nums: number[]): number[][] {
  const n: number = nums.length;
  const path: boolean[] = new Array(n).fill(false);
  const ans: number[][] = [];

  const backtrack = (prem: number[]) => {
    if (prem.length === n) {
      ans.push([...prem]);
      return;
    }

    for (let i = 0; i < n; i++) {
      if (path[i]) continue;
      path[i] = true;
      prem.push(nums[i]);
      backtrack(prem);
      path[i] = false;
      prem.pop();
    }
  };

  backtrack([]);
  return ans;
}

// Time:
// generate n! premutations
// each time you hit a full permutation, you copy perm into ans: O(n)
// O(n*n!)

// Space:
// O(n*n!)
